[CORDA-2923] - Make the default RPC client reconnect

dimosr opened the pull request:
This change makes the RPC client to automatically reconnect by default. It makes use of the existing `ReconnectingCordaRPCOps` functionality and the associated components.

Notes: 
- Reconnecting is disabled in the driver to prevent the rpc from trying to infinitely reconnect when the driver shuts down the nodes in the end of the test.
- The `ReconnectingObservableImpl ` was adjusted so that it uses the hook provided by the RxJava `Observables`. This is so that it works transparently for clients that can use the Observable APIs, i.e. `subscribe` etc.
- We will need to verify that the default client load balances between multiple addresses, when available. If that's true, then this will be maintained, otherwise we might need to add this enhancement. I'll discuss with Bogdan when he's back and perform a test on this.

tudor-malene commented on client/rpc/src/integration-test/kotlin/net/corda/client/rpc/CordaRPCClientReconnectionTest.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    @Test
+    fun `observables returned by the rpc client continue working when the server crashes and restarts`() {
```
the previous test can easily be merged in this one.

dimosr replied:
Yeah, in general I prefer separating tests that test different aspects. But, given that these tests are quite slow, I agree. I'll merge them.

tudor-malene commented on node/src/integration-test/kotlin/net/corda/node/services/rpc/RpcReconnectTests.kt:
```diff
@@ -283,8 +283,8 @@ class RpcReconnectTests {
-            vaultObserverHandle.stop()
-            stateMachineObserverHandle.stop()
+            vaultObserverHandle.unsubscribe()
+            stateMachineObserverHandle.unsubscribe()
```
Does this actually work?

dimosr replied:
Indeed, I don't think that works. Actually, I was planning to switch this test back to casting and using the observer handles to exercise the full functionality.

But, we will probably need to think a bit more about the unsubscription using the default APIs, because I think it won't work anymore.

mnesbit commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -382,10 +429,20 @@ class CordaRPCClient private constructor(
-    fun start(username: String, password: String, externalTrace: Trace?, impersonatedActor: Actor?, targetLegalIdentity: CordaX500Name?): CordaRPCConnection {
-        return CordaRPCConnection(getRpcClient().start(CordaRPCOps::class.java, username, password, externalTrace, impersonatedActor, targetLegalIdentity))
+    @JvmOverloads
+    fun start(username: String, password: String, externalTrace: Trace?, impersonatedActor: Actor?, targetLegalIdentity: CordaX500Name?, reconnecting: Boolean = true): CordaRPCConnection {
```
The interaction of this with HA is not at all clear. Does it correctly round-robin? What happens with a single RPC target (artemis is annoyingly inconsistent itself in handling failure). Please TEST this before this code goes any further.

tudor-malene replied:
Good point @mnesbit !
This hasn't been tested with HA at all.

If we're making this the default we'll have to test.

dimosr replied:
The `ActiveMQClient` is responsible for performing the round-robin on failover as long as it's provided with a list of all the available addresses. That's why I populate that conditionally on `hostAndPort` or `haAddressPool`.

We definitely need to test it works though. Bogdan gave me some pointers on how to programmatically setup an HA env - I will try to write an automated test to verify it works as expected during failover

tudor-malene replied:
Ideally you'd be able to upgrade the `RPCReconnectTest` on enterprise to create 2 bankA nodes in HA and kill them/their proxies randomly

dimosr commented:
> There's changes to the public API in CordaRPCClient.kt, which there can't be if this is targeted for 4.2. You will need to make workaround that, probably by making the new methods private/internal.

[Discussed offline] We'll keep it as-is for `master` and I will pay attention to introduce private methods when backporting it for 4.2

shamsasari commented:
Let me know when the testing is done and ready for review.

dimosr commented:
Added some tests for HA configuration and switched the default to be non-reconnecting because of some remaining issues with observables. I'd like to spend some more time to see if there's a way to intercept the unsubscription calls to the observables.

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -29,7 +30,47 @@ import java.util.ServiceLoader
+
+    private var reconnectingCordaRPCOps: ReconnectingCordaRPCOps? = null
+
+    constructor(connection: RPCConnection<CordaRPCOps>, username: String, password: String, addresses: List<NetworkHostAndPort>): this(connection) {
```
It seems the intention for the c'tors of this class are they be internal, even though in Java `internal` c'tors are still public. I guess that ship has sailed for the primary c'tor above, but there's no reason why we should continue on this path. Can you instead of this c'tor have an `internal` factory method annotated with `CordaInternal`.

dimosr replied:
definitely, I'll look into it.

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -382,10 +429,20 @@ class CordaRPCClient private constructor(
-        return CordaRPCConnection(getRpcClient().start(CordaRPCOps::class.java, username, password, externalTrace, impersonatedActor, targetLegalIdentity))
+    @JvmOverloads
+    fun start(username: String, password: String, externalTrace: Trace?, impersonatedActor: Actor?, targetLegalIdentity: CordaX500Name?, reconnecting: Boolean = false): CordaRPCConnection {
+        val addresses = if (haAddressPool.isEmpty())
```
Missing braces for the if-else, or more the entire expression onto a single line

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/ReconnectingCordaRPCOps.kt:
```diff
@@ -255,55 +255,66 @@ class ReconnectingCordaRPCOps private constructor(
+                                        val initial: DataFeed<*, T>,
+                                        val createDataFeed: () -> DataFeed<*, T>): OnSubscribe<T>, ReconnectingObservable<T> {
+            override fun call(child: rx.Subscriber<in T>?) {
+                subscribe( {t -> child!!.onNext(t) }, {}, {}, {})
```
You can simplify to just `child::onNext` if you remove the nullability of the param, which I don't think it is.

dimosr replied:
good catch, it was Intellij's autocomplete, I'll amend.

shamsasari commented on client/rpc/src/integration-test/kotlin/net/corda/client/rpc/CordaRPCClientReconnectionTest.kt:
```diff
@@ -0,0 +1,96 @@
+import java.util.concurrent.TimeUnit
+import kotlin.test.assertTrue
+
+class CordaRPCClientReconnectionTest {
```
What's the difference between this and RpcReconnectTests?

dimosr replied:
I think these are closer to functional tests, while the other ones are closer to stress/chaos testing. For example, these are deterministic and faster to execute, while the other ones are probabilistic and a bit slower.

dimosr replied:
I was thinking of renaming them btw and adding a comment to clarify that, but forgot it. I'll do that now.

shamsasari commented:
Can you quickly describe how the API works, because I don't like the idea of exposing an reconnecting abstraction to work around limitations with RxJava. It seems like we have two ways for listening for the same thing. 

The original ticket references suggestions from Mike on how to re-write RPCs and observations and introduction of a new Idempotent annotation. What became of that?

dimosr commented:
@shamsasari I was intentionally delaying the docs update until we have a firm decision on the changes from a client perspective. I am planning on updating this part https://docs.corda.net/clientrpc.html?highlight=rpc#reconnecting-rpc-clients

Based on the current state, the API would work as follows:
- if the rpc client is instantiated in the default way (`reconnecting` = `false`), everything works as before exactly.
- if the rpc client is instantiated using `reconnecting` = `false`, then
  - the client reconnects under the hood. All the APIs will now block and retry in the case of a node outage, instead of throwing an exception.
  - the APIs returning streams will still return `Observable`. Subscription on these observables will work  and subscribers will keep receiving updates between reconnections transparently. However, unsubscriptions to the observables will not work (due to the RxJava limitations), so the subscribers will keep getting updates. In order to unsubscribe successfully, the client will have to execute `asReconnecting()` to the returned observable in order to get a `ReconnectingObservable` and then will be able to both subscribe and unsubscribe.

dimosr commented:
[Note to myself]: add another converter method along the extension function `asReconnecting()` for Java clients.

tudor-malene commented:
> [Note to myself]: add another converter method along the extension function `asReconnecting()` for Java clients.

Java clients can just cast the old fashioned way.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -29,7 +30,47 @@ import java.util.ServiceLoader
+        this.reconnectingCordaRPCOps = ReconnectingCordaRPCOps(addresses, username, password)
+    }
+
+    private val actualConnection: RPCConnection<CordaRPCOps> by lazy {
```
I didn't like it either, but there was a circular reference that was causing infinite recursion at some point. Thinking it twice, there's no really any reason for a member variable, I'll just make it a function.

dimosr commented:
> This is missing documentation, and so I can't easily see how this API works. I believe the new API is ReconnectingObservable? If so then it's still in an internal package.

@shamsasari regarding this, I was thinking of moving only the `ReconnectingObservable`, `ReconnectingObservableImpl` and `asReconnecting()` converter method in a publicly visible package and keep the rest as-is. Let me know if you disagree.

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -382,10 +427,22 @@ class CordaRPCClient private constructor(
+        val addresses = if (haAddressPool.isEmpty()) {
+            listOf(hostAndPort!!)
+        }
+        else {
```
Move `else` up

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -382,10 +427,22 @@ class CordaRPCClient private constructor(
+            haAddressPool
+        }
+
+        return if (reconnecting)
```
braces for if-else

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -334,10 +373,12 @@ class CordaRPCClient private constructor(
-    fun start(username: String, password: String): CordaRPCConnection {
-        return start(username, password, null, null)
+    @JvmOverloads
+    fun start(username: String, password: String, reconnecting: Boolean = false): CordaRPCConnection {
```
`reconnecting` -> `autoReconnect`

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/reconnect/ReconnectingObservable.kt:
```diff
@@ -0,0 +1,93 @@
+import rx.Observable
+import java.util.concurrent.ExecutorService
+
+class ReconnectingObservableImpl<T> internal constructor(
```
This needs to be in the internal package

shamsasari commented:
You need to rebase against master as well.

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/ReconnectingObservableImpl.kt:
```diff
@@ -0,0 +1,75 @@
+                                    val initial: DataFeed<*, T>,
+                                    val createDataFeed: () -> DataFeed<*, T>): OnSubscribe<T>, ReconnectingObservable<T> {
+        override fun call(child: rx.Subscriber<in T>) {
+            subscribe( {t -> child.onNext(t) }, {}, {}, {})
```
`subscribe(child::onNext, ...`

shamsasari commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/ReconnectingObservableImpl.kt:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        private var initialStartWith: Iterable<T>? = null
+        fun _subscribeWithReconnect(observerHandle: ObserverHandle, onNext: (T) -> Unit, onStop: () -> Unit, onDisconnect: () -> Unit, onReconnect: () -> Unit, startWithValues: Iterable<T>? = null) {
```
Intellij is saying this can be private

shamsasari replied:
Can we also give rid of the underscore

shamsasari commented:
`ReconnectingCordaRPCOps.runFlowAndReturnResultWithLogicalRetry` is not used and can be deleted.

dimosr commented:
@shamsasari I updated the docs, have a look and let me know if something's not clear. I am looking on the unsubscription change we discussed in the meanwhile.

wrt.
> ReconnectingCordaRPCOps.runFlowAndReturnResultWithLogicalRetry is not used and can be deleted.

I don't have a lot of context on this since it predates these changes - @tudor-malene can we remove this or is it still needed ?

tudor-malene commented:
> > ReconnectingCordaRPCOps.runFlowAndReturnResultWithLogicalRetry is not used and can be deleted.
> 
> I don't have a lot of context on this since it predates these changes - @tudor-malene can we remove this or is it still needed ?

It was thought as a utility for people who want tot do that (run a flow reliably and get the result).
It should be tested though.

dimosr commented:
> It was thought as a utility for people who want tot do that (run a flow reliably and get the result).
It should be tested though.

Thanks for the context. I think testing that is out-of-context for this change, so I am inclined to remove it as well given it's not part of the public API yet and not mentioned in the related docs.

tudor-malene commented:
> > It was thought as a utility for people who want tot do that (run a flow reliably and get the result).
> > It should be tested though.
> 
> Thanks for the context. I think testing that is out-of-context for this change, so I am inclined to remove it as well given it's not part of the public API yet and not mentioned in the related docs.

That's ok, but maybe create a ticket to add and test this in the future.
It is 100% something that people will do

dimosr commented:
sure: https://r3-cev.atlassian.net/browse/CORDA-3049

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -30,7 +32,44 @@ import java.util.ServiceLoader
  * @see RPCConnection
  */
-class CordaRPCConnection internal constructor(connection: RPCConnection<CordaRPCOps>) : RPCConnection<CordaRPCOps> by connection
+class CordaRPCConnection internal constructor(private val connection: RPCConnection<CordaRPCOps>) : RPCConnection<CordaRPCOps> {
```
None really, I just avoided to change the constructor API because it was public. The nullability though does not leak in the ABI, so I can make it nullable safely I believe.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -30,7 +32,44 @@ import java.util.ServiceLoader
+        }
+    }
+
+    private var reconnectingCordaRPCOps: ReconnectingCordaRPCOps? = null
```
If we had constructors, I believe yes, but now I don't think so, since it's considered as re-assigned. I gave it a try, but the static type system complains.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -30,7 +32,44 @@ import java.util.ServiceLoader
+        @CordaInternal
+        internal fun createWithReconnection(connection: RPCConnection<CordaRPCOps>, username: String, password: String, addresses: List<NetworkHostAndPort>): CordaRPCConnection {
+            val cordaRpcConnection = CordaRPCConnection(connection)
+            cordaRpcConnection.reconnectingCordaRPCOps = ReconnectingCordaRPCOps(addresses, username, password)
```
nothing really - related to your comment above. Making it nullable, the factory method could then just pass `null` in to make this more clear.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/ObserverHandle.kt:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Handle to externally control a subscribed observer for [ReconnectingObservable]s.
+ */
+class ObserverHandle {
```
do you mean the component ? It's still needed as it's used from `ReconnectingObservableImpl` to synchronize on errors, reconnections etc.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/ReconnectingObservable.kt:
```diff
@@ -0,0 +1,14 @@
+ *
+ * TODO - provide a logical function to know how to retrieve missing events that happened during disconnects.
+ */
+interface ReconnectingObservable<T> {
```
I think it depends on the definition of "need". The methods defined in this interface are still used by some internal components. Arguably, we could remove it and use directly the `ReconnectingObservableImpl`, but I think it provides a good abstraction and it's worth keeping it.

shamsasari replied:
Oh I didn't realised this was internal.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt:
```diff
@@ -350,10 +391,12 @@ class CordaRPCClient private constructor(
      * @param username The username to authenticate with.
      * @param password The password to authenticate with.
      * @param targetLegalIdentity in case of multi-identity RPC endpoint specific legal identity to which the calls must be addressed.
+     * @param autoReconnect whether the connection will try to reconnect automatically
```
sure, let me give it a try.

shamsasari commented on client/rpc/src/integration-test/kotlin/net/corda/client/rpc/CordaRPCClientReconnectionTest.kt:
```diff
@@ -0,0 +1,124 @@
+            val networkParameters = rpcOps.networkParameters
+            val cashStatesFeed = rpcOps.vaultTrack(Cash.State::class.java)
+            cashStatesFeed.updates.subscribe { latch.countDown() }
+            rpcOps.startTrackedFlow(::CashIssueFlow, 10.DOLLARS, OpaqueBytes.of(0), defaultNotaryIdentity).returnValue.get()
```
Why do we have this issue flow here? Surely to test the latch is counted down after the restart we should only issue the flow after the restart? Commenting this line out fails the test. Or am I missing something here?

dimosr replied:
If I understand correctly, commenting out the line is expected to make the test fail, since the assertion waits for the latch to receive 2 count downs and will only receive one in that case.

You are right - we could remove this issuance and also adjust the invariance check to assert against 1, instead of 2. I wrote it this way, because it felt more natural to me and easier to understand for a newcomer to the code - "trigger 2 flows one before the crash and one after, expect that the initial rpc client will observe both.". Happy to change if you feel it's not for the best.

shamsasari replied:
Ah, the latch starts at 2. Gotcha! No, what you have it fine. Maybe rename to make it more clear.

shamsasari commented:
I've removed the ``ReconnectingObservable`` interface and renamed ``ReconnectingObservableImpl`` to it. Also the re-connect handling has been simplified and no longer requires the ``ObserverHandle``.

I've also removed any mention to the internal sample reconnecting code we had. There's no longer any need with this new public API.

tudor-malene commented:
> I've removed the `ReconnectingObservable` interface and renamed `ReconnectingObservableImpl` to it. Also the re-connect handling has been simplified and no longer requires the `ObserverHandle`.
> 
> I've also removed any mention to the internal sample reconnecting code we had. There's no longer any need with this new public API.

Nice work!
As mentioned the  `onDisconnect`/`onReconnect` hooks are necessary for UIs like the node explorer.
Also, worth playing a bit with the explorer, to check if it works properly. (I'm not 100% sure if the initialValues work )

dimosr commented:
@tudor-malene I had an offline chat with Shams and we agreed on the following: the handlers for reconnection are indeed useful from a UI perspective, but it would be better to have them at the level where the client is initialised, since they are mostly related to the client and not the observables. So, we decided to remove them for now to bring this to a close and create a separate ticket to follow-up on adding them in the right place. See: https://r3-cev.atlassian.net/browse/CORDA-3085 (@shamsasari feel free to do the necessary admin).

With regards to the Explorer at the moment, I did a test and the current behaviour is the following: if one performs an action (i.e. issue/pay) during a server restart, then the loading spinner appears (as in the default case) and disappears as soon as the connection is re-established with the appropriate message having executed the transaction successfully.

tudor-malene commented:
> @tudor-malene I had an offline chat with Shams and we agreed on the following: the handlers for reconnection are indeed useful from a UI perspective, but it would be better to have them at the level where the client is initialised, since they are mostly related to the client and not the observables. So, we decided to remove them for now to bring this to a close and create a separate ticket to follow-up on adding them in the right place. See: https://r3-cev.atlassian.net/browse/CORDA-3085 (@shamsasari feel free to do the necessary admin).

@thschroeter is using `onReconnect` for the vault reconciliation .
I would recommend having a chat with him as well.

shamsasari commented:
Thomas' work is currently targeted for ENT 3, so let's not block this on that. We just need to make sure the API design for reconnect hooks is correct, hence the separate ticket.
